### DISCUSSION TITLE
Limit of triangles in MeshShape?

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#411] [Q&A] [answered]
I'm guessing that the tree used by this shape has some kind of limit - can it be calculated?
For context, at the moment I'm debugging a problem (a crash) that may be caused by a mesh with almost 24k triangles 😅
This mesh triggers this assertion in SubShapeIDCreator::PushID()
JPH_ASSERT(copy.mCurrentBit <= SubShapeID::MaxBits);

**Comment by jrouwe** [comment]
There is a limit (I don't know the exact limit by heart as it depends on the tree structure that is built). If you exceeded the limits then it would have failed creation. 24k triangles should definitively fit.

The SubShapeIDs indicate a path from body to the leaf shape (triangle in this case). My guess is that your mesh is in e.g. a StaticCompoundShape with lots of children (this will also eat bits from your SubShapeID) or in multiple levels of StaticCompoundShapes. Alternatively you could simply be passing an incorrect SubShapeID which would crash the system as it tries to follow the path to the leaf.

What's the callstack of the assert?

**Reply by DocAce** [reply]
Oh yeah! It's in a CompoundShape with... almost 7k child shapes 😬
That's what you get if you let players expand their space stations without limitations 😅

Here's the callstack for the assert:

```
JPH::SubShapeIDCreator::PushID(unsigned int) Line 104
JPH::MeshShape::WalkTreePerTriangle::__l2::ChainedVisitor::VisitTriangles(const JPH::TriangleCodecIndexed8BitPackSOA4Flags::DecodingContext &) Line 479
JPH::NodeCodecQuadTreeHalfFloat<1>::DecodingContext::WalkTree(const unsigned char *) Line 259
JPH::MeshShape::WalkTree(JPH::MeshShape::WalkTreePerTriangle::__l2::ChainedVisitor &) Line 431
JPH::MeshShape::WalkTreePerTriangle(const JPH::SubShapeIDCreator &) Line 495
JPH::MeshShape::CastRay(const JPH::RayCast & inRay, const JPH::RayCastSettings & inRayCastSettings, const JPH::SubShapeIDCreator & inSubShapeIDCreator, JPH::CollisionCollector<JPH::RayCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::ShapeFilter & inShapeFilter) Line 768
XPhys::XWrapperShape::CastRay(const JPH::RayCast & inRay, const JPH::RayCastSettings & inRayCastSettings, const JPH::SubShapeIDCreator & inSubShapeIDCreator, JPH::CollisionCollector<JPH::RayCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::ShapeFilter & inShapeFilter) Line 562
JPH::CompoundShape::CastRayVisitorCollector::VisitShape(const JPH::CompoundShape::SubShape &) Line 99
JPH::MutableCompoundShape::WalkSubShapes<`JPH::MutableCompoundShape::CastRay'::`2'::Visitor>(JPH::MutableCompoundShape::CastRay::__l2::Visitor & ioVisitor) Line 264
JPH::MutableCompoundShape::CastRay(const JPH::RayCast & inRay, const JPH::RayCastSettings & inRayCastSettings, const JPH::SubShapeIDCreator & inSubShapeIDCreator, JPH::CollisionCollector<JPH::RayCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::ShapeFilter & inShapeFilter) Line 335
JPH::TransformedShape::CastRay(const JPH::RayCast & inRay, const JPH::RayCastSettings & inRayCastSettings, JPH::CollisionCollector<JPH::RayCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::ShapeFilter & inShapeFilter) Line 60
JPH::NarrowPhaseQuery::CastRay'::`2'::MyCollector::AddHit(const JPH::BroadPhaseCastResult & inResult) Line 128
JPH::QuadTree::CastRay::__l2::Visitor::VisitBody(const JPH::BodyID &) Line 1078
JPH::QuadTree::WalkTree(const JPH::ObjectLayerFilter &) Line 971
JPH::QuadTree::CastRay(const JPH::RayCast & inRay, JPH::CollisionCollector<JPH::BroadPhaseCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::ObjectLayerFilter & inObjectLayerFilter, const std::vector<JPH::QuadTree::Tracking,JPH::STLAllocator<JPH::QuadTree::Tracking>> & inTracking) Line 1089
JPH::BroadPhaseQuadTree::CastRay(const JPH::RayCast & inRay, JPH::CollisionCollector<JPH::BroadPhaseCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::BroadPhaseLayerFilter & inBroadPhaseLayerFilter, const JPH::ObjectLayerFilter & inObjectLayerFilter) Line 411
JPH::NarrowPhaseQuery::CastRay(const JPH::RayCast & inRay, const JPH::RayCastSettings & inRayCastSettings, JPH::CollisionCollector<JPH::RayCastResult,JPH::CollisionCollectorTraitsCastRay> & ioCollector, const JPH::BroadPhaseLayerFilter & inBroadPhaseLayerFilter, const JPH::ObjectLayerFilter & inObjectLayerFilter, const JPH::BodyFilter & inBodyFilter, const JPH::ShapeFilter & inShapeFilter) Line 144
XPhys::CastRay(XPhys::XCastRayCollectorBase & collector, const XPhys::SysWorld * sworld, const JPH::RayCast & raycast, XPhys::Body * targetbody, XLib::UncountedPtr<U::Component const> targetcomponent, bool includechildren, XPhys::Body * excludedbody, XLib::UncountedPtr<U::Component const> excludedcomponent, bool excludechildren, XLib::TagID excludedtag, XPhys::CollisionGroup group, XPhys::CollisionMask mask, bool highprecision) Line 1277
```

(XPhys::XWrapperShape::CastRay() does nothing except pass directly to mInnerShape->CastRay(). It's mostly used to map compound children to our game objects, and the userdata isn't large enough for that.)

Is there anything I can do here? Could there be a way to "reset" the sub shape ID when it gets to the MeshShape?

**Comment by jrouwe** [comment, marked as answer]
7k child shapes eats up 13 out of 32 bits for the SubShapeID (see CompoundShape::GetSubShapeIDBits), so if your mesh uses 19 or more bits then it won't fit. You can call Shape::GetSubShapeIDBitsRecursive to find out. MutableCompoundShape checks this at construction time, but if you add additional shapes later there's no validation.

The solution is to check the bits needed before creating a compound and create multiple bodies if it does not fit (e.g. have 2 bodies, each with a compound shape of 3.5k child shapes). I'm assuming we're talking about a static body here b.t.w.

B.t.w. a MutableCompoundShape with 7k children is very inefficient because it doesn't have an internal tree structure. I would change it to a StaticCompoundShape and create a new shape when it needs to change (hopefully not every frame).

**Reply by DocAce** [reply]
Thanks!
Unfortunately almost all of our objects have dynamic parts, but if I'm going to end up separating things into multiple bodies anyway I may as well split them up by that and have some bodies with a StaticCompoundShape.
